### PR TITLE
Add `from_shamir_mnemonic_phrases` and `generate_shamir_mnemonic_phrases` to `Keypair`

### LINKED ISSUE #1010: Add Shamir's secret sharing protocol
**Is your feature request related to a problem? Please describe.**

Keeping secrets can be hard and if someone finds the paper wallet, it's compromised.

**Describe the solution you'd like**

I would like to propose to add a helper function to split the secret key using the Shamir protocol. The PyCryptodome library has an implementation we could leverage.

We could propose the following APIs:
```python
Keypair.secret.to_shamir(n, k)
Keypair.from_shamir(...)
```

Of course, naming and input parameters should be worked on.

As usual, I am happy to make a PR if wanted 😃 

**Describe alternatives you've considered**

If not suitable in this library, I will add the feature on my higher level Soroban API library.

**Additional context**

https://pycryptodome.readthedocs.io/en/latest/src/protocol/ss.html

### PR DESCRIPTION
**Pull request by tupui**
Closes #1010

Add the following API to `Keypair`
```python
from stellar_sdk import Keypair
mnemonic_phrase = Keypair.generate_shamir_mnemonic_phrases(2, 3)
Keypair.from_shamir_mnemonic_phrases(mnemonic_phrase[1:])
<Keypair [public_key=GD5BVWNCAIEHZL6X2XOYWFLHIZVX4MZVDHWGZCC5K2YSQ6N2ZGE25KP4, private_key_exists=True]>
```

I will can add more docs and tests once we settle on the API.

I was unsure if you wanted to add a utility to also apply Shamir on an existing mnemonic phrase or secret, and if yes, if that should be part of the same generate method.

Happy to adjust anything 😃

### CODE REVIEW

File: stellar_sdk/keypair.py [hunk 1]
```diff
@@ -256,6 +258,54 @@ def from_mnemonic_phrase(
...
+        :param passphrase: An optional passphrase used to decrypt the secret key.
+        :return: A new :class:`Keypair` object derived from the mnemonic phrases.
+        """
+        try:
+            secret = shamir_mnemonic.combine_mnemonics(
+                mnemonics=mnemonic_phrases, passphrase=passphrase.encode()
+            )
+        except shamir_mnemonic.utils.MnemonicError as exc:
+            raise ValueError(exc) from exc
+        return cls.from_secret(secret.decode())
```
Comment (overcat) [inline stellar_sdk/keypair.py:307]: What you have here is a master seed, and you cannot directly generate a KeyPair from it. Please refer to [SLIP-10](https://github.com/satoshilabs/slips/blob/master/slip-0010.md) for more information. 

Here is a fully functional example.
```python
import shamir_mnemonic
from stellar_sdk import Keypair

from stellar_sdk.sep.mnemonic import StellarMnemonic

# 2-of-3 Shamir Secret Sharing Scheme, generate from Trezor Safe 3
share0 = "glimpse buyer academic acid branch sled disaster sunlight material junction float emperor intend priority scene trash remember radar prospect dryer"
share1 = "glimpse buyer academic agency burden payroll alpha oven large amount smear forward pharmacy symbolic junk axle exercise segment frequent axle"
share2 = "glimpse buyer academic always careful become dance teaspoon daisy orange careful steady boundary exceed robin remind software grin space advocate"

master_seed = shamir_mnemonic.combine_mnemonics([share0, share2], passphrase=b"9012")

kp0 = Keypair.from_raw_ed25519_seed(StellarMnemonic.derive(master_seed, 0))  # m/44h/148h/0h
kp1 = Keypair.from_raw_ed25519_seed(StellarMnemonic.derive(master_seed, 1))
kp100 = Keypair.from_raw_ed25519_seed(StellarMnemonic.derive(master_seed, 100))

assert kp0.public_key == "GDZ4GYLVRLM2E6CGCOVYXAYMXJJAV3IHDXU6RUHX5AJVYS4AE6R6CHPJ"
assert kp1.public_key == "GCMZKXAAPQ3TDY5P7QDUVJBW66R2DGT7AM6MA3MCQENIF37E25U2PEK3"
assert kp100.public_key == "GCF7DAVTXXVQPOSB5TCA2CIFT7DZIPK23NCOV3RJ6FTYTZM6S6RPPACM"

```

File: stellar_sdk/keypair.py [hunk 2]
```diff
@@ -256,6 +258,54 @@ def from_mnemonic_phrase(
...
+        :return: A list of mnemonic phrases.
+        """
+        if member_count < member_threshold:
+            raise ValueError(
+                "There must be more members than the set threshold: "
+                "``member_count > member_threshold``."
+            )
+        secrets = Keypair.random().secret.encode()
+        return shamir_mnemonic.generate_mnemonics(
+            group_threshold=1,
```
Comment (overcat) [inline stellar_sdk/keypair.py:283]: Only supporting a single group should be sufficient to meet the needs of the vast majority of users.

**Review by overcat — CHANGES_REQUESTED** [review]
> I was unsure if you wanted to add a utility to also apply Shamir on an existing mnemonic phrase or secret, and if yes, if that should be part of the same generate method.

I tend not to support it because the master seed will change, and the resulting KeyPair will also change.

File: stellar_sdk/keypair.py (hunk 1 above)
Comment (tupui) [reply stellar_sdk/keypair.py]: Ok, I have added the index support. I can use what you have here for a test afterwards.

Let me know if that's what you had in mind.

File: pyproject.toml [hunk 3]
```diff
@@ -53,6 +53,7 @@ requests = "^2.32.3"
 aiohttp = { version = "^3.9.1", optional = true}
 aiohttp-sse-client = { version = "^0.2.1", optional = true}
 mnemonic = "^0.20"
+shamir-mnemonic = "^0.3.0"
```
Comment (overcat) [inline pyproject.toml:56]: I suggest setting it as an optional dependency, as SLIP 39 does not seem to be widely used yet.
```
shamir-mnemonic = { version = "^0.3.0", optional = true }

[tool.poetry.extras]
aiohttp = ["aiohttp", "aiohttp-sse-client"]
shamir = ["shamir-mnemonic"]
```

**Review by overcat — COMMENTED** [review]
It looks great; we just need to add a few more tests.

**Comment by tupui** [comment]
> It looks great; we just need to add a few more tests.

Ok great I will do that when I have some time and also add some docs.

**Comment by tupui** [comment]
@overcat I have integrated your suggestions and added tests and docs.

---

Otherwise, I was thinking that we could actually consider merging the two mnemonic ways `from_shamir_mnemonic_phrases` and `from_mnemonic_phrase` (resp. `generate_xxx`). In the end there is not much difference in the API and it's just a matter of asking for a split of not. You can do `generate_shamir_mnemonic_phrases(1, 1)` which is equivalent to `generate_mnemonic_phrase()`. Then the reconstruction can go to one or the other implementation depending on the type of the arg. The main issue I see is with the support of  languages and strength in `generate_mnemonic_phrase`.

File: pyproject.toml (hunk 3 above)
Comment (tupui) [reply pyproject.toml]: I added the import in each function with a check. This way there is no global import. I can also use a try except block and a `HAS_SHAMIR` variable if you prefer that.

Comment (overcat) [reply pyproject.toml]: 👍 Your approach is great and doesn't need any changes.

**Comment by overcat** [comment]
> @overcat I have integrated your suggestions and added tests and docs.
> 
> Otherwise, I was thinking that we could actually consider merging the two mnemonic ways `from_shamir_mnemonic_phrases` and `from_mnemonic_phrase` (resp. `generate_xxx`). In the end there is not much difference in the API and it's just a matter of asking for a split of not. You can do `generate_shamir_mnemonic_phrases(1, 1)` which is equivalent to `generate_mnemonic_phrase()`. Then the reconstruction can go to one or the other implementation depending on the type of the arg. The main issue I see is with the support of languages and strength in `generate_mnemonic_phrase`.

Hi @tupui, let's maintain this state for now, after all, these are two different standards.

**Comment by tupui** [comment]
Thanks for getting this in 😃🙌

**Comment by tomerweller** [comment]
Very cool. Should we standardize shamir secret sharing in stellar ecosystem as a SEP?

**Comment by overcat** [comment]
> Very cool. Should we standardize shamir secret sharing in stellar ecosystem as a SEP?

I think we just need to revise SEP-05: add the part related to SLIP-39 in it.

**Comment by tupui** [comment]
Happy to help update the SEP

**Comment by overcat** [comment]
Hi @tupui, if you are interested, you can submit a PR for [this document](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0005.md). Feel free to ping me if you need any assistance.

**Comment by tupui** [comment]
Thanks I will do 👍 might just take some days as I am on holiday.

**Comment by tupui** [comment]
@overcat @tomerweller I had a closer look and I am not sure about something.

For the seed, in SEP-5 it is said that mnemonic should be used to get a BIP39 seed (the SLIP-39, which deals about Shamir, also points to use a BIP39 seed.) But in the Python SDK to me it looks like we are using instead an ED25519.

---

Some Python code I am writing to get this right and generate the data for the SEP update.

```python
import shamir_mnemonic
from stellar_sdk import Keypair
from stellar_sdk.sep.mnemonic import StellarMnemonic
from mnemonic import Mnemonic


# from SEP-5 test cases
mnemonic = "illness spike retreat truth genius clock brain pass fit cave bargain toe"
bip39_seed = "e4a5a632e70943ae7f07659df1332160937fad82587216a4c64315a0fb39497ee4a01f76ddab4cba68147977f3a147b6ad584c41808e8238a07f6cc4b582f186"
passphrase = ""
language = "english"

kp = Keypair.from_mnemonic_phrase(mnemonic_phrase=mnemonic)
raw_ed25519_seed = StellarMnemonic(language).to_seed(mnemonic, passphrase)

# if we wanted to use a bip39 seed instead
seed = Mnemonic("english").to_seed(mnemonic).hex()
assert seed == bip39_seed

shamir = shamir_mnemonic.generate_mnemonics(
    group_threshold=1,
    groups=[(2, 3)],
    master_secret=raw_ed25519_seed,
    passphrase=passphrase.encode(),
)[0]

reconstructed_raw_ed25519_seed = shamir_mnemonic.combine_mnemonics(
    mnemonics=shamir[:2], passphrase=passphrase.encode()
)

derived_seed = StellarMnemonic.derive(seed=reconstructed_raw_ed25519_seed, index=0)
reconstructed_kp = Keypair.from_raw_ed25519_seed(reconstructed_raw_ed25519_seed)

assert reconstructed_kp.public_key() == kp.public_key()
```

**Comment by overcat** [comment]
Hi @tupui, 

The complete process is as follows: BIP39 seed -> SLIP10 master key -> SLIP10 child key -> ED25519 keypair

`stretched[:64]` is the BIP39 seed, and then we generate the SLIP10 child key through `derive`, which is used as the seed for generating the ED25519 keypair.
Please check https://github.com/StellarCN/py-stellar-base/blob/94b7417fe409a1c3c33251e32081a2d605361bb2/stellar_sdk/sep/mnemonic.py#L63

**Comment by tupui** [comment]
Ok 👍 

For the generate function it does not matter much, though we should maybe instead use `os.urandom` instead of using the secret key as the secret to be encoded by Shamir should be a BIP39.

For `from_shamir_mnemonic_phrases` I think there is can be an issue because `shamir_mnemonic.combine_mnemonics` returns a BIP39 (if we generated the phrases with a BIP39 seed as per the SLIP-39). So we should convert this seed to an ED25519 before passing it to `StellarMnemonic.derive`.

If we don't change anything, then it's the code above where we need to generate the Shamir phrases with the ED25519. But that means someone converting a mnemonic to Shamir needs to know that otherwise it won't be consistent with the reconstruction we do here.

**Comment by overcat** [comment]
Hi @tupui,
> For the generate function it does not matter much, though we should maybe instead use a version of `os.urandom` instead of using the secret key as the secret to be encoded by Shamir should be a BIP39.

The current implementation, aside from not being able to set the `seed` length, has no other impact; even if we add a `strength` parameter in the future and switch to `secrets.token_bytes` or `os.urandom`, it will not result in breaking changes.

> For from_shamir_mnemonic_phrases I think there is can be an issue because shamir_mnemonic.combine_mnemonics returns a BIP39 (if we generated the phrases with a BIP39 seed as per the SLIP-39). So we should convert this seed to an ED25519 before passing it to StellarMnemonic.derive.

For SLIP-39, the process is SLIP39 seed -> SLIP10 master key -> SLIP10 child key -> ED25519 keypair. Currently, our code fully complies with this process, and I believe we do not need any modifications. `StellarMnemonic.derive` here accepts the SLIP-39 seed.

However, we can move `StellarMnemonic.derive` to a separate file, such as slip10.py, to avoid misunderstandings.

**Comment by tupui** [comment]
To clarify, my understanding with what you said means that the following should work. 

I must be missing something there.

```python

import shamir_mnemonic
from stellar_sdk import Keypair
from stellar_sdk.sep.mnemonic import StellarMnemonic
from mnemonic import Mnemonic


# from SEP-5 test cases
mnemonic = "illness spike retreat truth genius clock brain pass fit cave bargain toe"
bip39_seed = "e4a5a632e70943ae7f07659df1332160937fad82587216a4c64315a0fb39497ee4a01f76ddab4cba68147977f3a147b6ad584c41808e8238a07f6cc4b582f186"
passphrase = ""
language = "english"

kp = Keypair.from_mnemonic_phrase(mnemonic_phrase=mnemonic)

# bip39 seed from mnemonic
seed = Mnemonic("english").to_seed(mnemonic).hex()
assert seed == bip39_seed

# Shamir with the bip39 seed
shamir = shamir_mnemonic.generate_mnemonics(
    group_threshold=1,
    groups=[(2, 3)],
    master_secret=seed,
    passphrase=passphrase.encode(),
)[0]

reconstructed_seed = shamir_mnemonic.combine_mnemonics(
    mnemonics=shamir[:2], passphrase=passphrase.encode()
)

# bip39 -> ... -> ED25519
derived_seed = StellarMnemonic.derive(seed=reconstructed_seed, index=0)

reconstructed_kp = Keypair.from_raw_ed25519_seed(derived_seed)

assert reconstructed_kp.public_key == kp.public_key
```

**Comment by tupui** [comment]
Seen in chat. The above code does work after converting the seed from hex to binary 😅
I will generate the test data and update the SEP. Once all that is done I can also make a PR here to add these here so the test suite is complete.

**Pull request merged**